src/tree/updater_quantile_hist.cc

```diff
@@ -1099,24 +1137,26 @@ void QuantileHistMaker::Builder::InitNewNode(int nid,
 
   {
     auto& stats = snode_[nid].stats;
-    GHistRow hist = hist_[nid];
+    GHistRow<GradientSumT> hist = hist_[nid];
+    xgboost::detail::GradientPairInternal<GradientSumT> grad_stat;
     if (tree[nid].IsRoot()) {
       if (data_layout_ == kDenseDataZeroBased || data_layout_ == kDenseDataOneBased) {
         const std::vector<uint32_t>& row_ptr = gmat.cut.Ptrs();
         const uint32_t ibegin = row_ptr[fid_least_bins_];
         const uint32_t iend = row_ptr[fid_least_bins_ + 1];
         auto begin = hist.data();
         for (uint32_t i = ibegin; i < iend; ++i) {
-          const GradStats et = begin[i];
-          stats.Add(et.sum_grad, et.sum_hess);
+          const xgboost::detail::GradientPairInternal<GradientSumT> et = begin[i];
+          grad_stat.Add(et.GetGrad(), et.GetHess());
         }
       } else {
         const RowSetCollection::Elem e = row_set_collection_[nid];
         for (const size_t* it = e.begin; it < e.end; ++it) {
-          stats.Add(gpair[*it]);
+          grad_stat.Add(gpair[*it].GetGrad(), gpair[*it].GetHess());
         }
       }
-      histred_.Allreduce(&snode_[nid].stats, 1);
+      histred_.Allreduce(&grad_stat, 1);
+      snode_[nid].stats = tree::GradStats(grad_stat.GetGrad(), grad_stat.GetHess());
```

Comment:
I suppose we can keep usage of double always here. It shouldn't affect the performance, computation will stay more stable and no additional code changes.

Reply:
this block of code can't be unmodified as hist.data() has different type now.
And `histred_.Allreduce` can't accept `stats`